In [1]:
import pandas as pd

df = pd.read_csv("../data/ecommerce_dataset.csv")

# EDA

In [2]:
df.head()

,order_id,order_date,customer_id,customer_age,gender,region,product_name,category,quantity,unit_price,discount_percent,order_amount,customer_segment,is_repeat_customer,delivery_days,customer_rating
0,ORD-100000,2024-08-16,CUST-1035,65.0,Male,Saint Petersburg,AirPods Pro,Headphones,2,250,5,475.0,Business,True,1.0,NaN
1,ORD-100001,2025-10-27,CUST-1179,52.0,Female,Novosibirsk,iPhone 15,Smartphones,2,900,25,1350.0,Business,False,1.0,4.5
2,ORD-100002,2024-08-08,CUST-1195,39.0,Male,Saint Petersburg,Google Pixel 9,Smartphones,1,800,5,760.0,Business,True,NaN,NaN
3,ORD-100003,2024-05-07,CUST-1096,23.0,Female,Krasnodar,AirPods Pro,Headphones,2,250,20,400.0,Standard,True,1.0,4.3
4,ORD-100004,2024-04-13,CUST-1097,35.0,Female,Krasnodar,PlayStation 5,Consoles,3,600,30,1260.0,Standard,False,6.0,3.4


In [3]:
df.shape

(508, 16)

Количество строк в датасете: 508.

Количество столбцов в датасете: 16.

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 508 entries, 0 to 507
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            508 non-null    str    
 1   order_date          508 non-null    str    
 2   customer_id         508 non-null    str    
 3   customer_age        483 non-null    float64
 4   gender              493 non-null    str    
 5   region              508 non-null    str    
 6   product_name        508 non-null    str    
 7   category            508 non-null    str    
 8   quantity            508 non-null    int64  
 9   unit_price          508 non-null    int64  
 10  discount_percent    508 non-null    int64  
 11  order_amount        508 non-null    float64
 12  customer_segment    508 non-null    str    
 13  is_repeat_customer  508 non-null    bool   
 14  delivery_days       473 non-null    float64
 15  customer_rating     457 non-null    float64
dtypes: bool(1), float64

Числовые признаки: типы int и float.

Категориальные признаки: str.

Булевые значения: bool.

Столбец order_date нужно будет привести к типу datetime.

In [5]:
df.isnull().sum()

order_id               0
order_date             0
customer_id            0
customer_age          25
gender                15
region                 0
product_name           0
category               0
quantity               0
unit_price             0
discount_percent       0
order_amount           0
customer_segment       0
is_repeat_customer     0
delivery_days         35
customer_rating       51
dtype: int64

Пропуски присутствуют в столбцах customer_age, gender, delivery_dates и customer_rating.

Наибольшее количество пропусков в столбце customer_rating.

Пропуски в столбцах customer_age и gender можно будет удалить.

Пропуски в столбцах delivery_dates и customer_rating, при необходимости, можно будет привести к среднему значению.

In [6]:
df.duplicated().sum()

np.int64(7)

В датасете обнаружено 7 дубликатов. 

В ходе анализа можно будет оценить необходимость их удаления.

In [7]:
df.describe()

,customer_age,quantity,unit_price,discount_percent,order_amount,delivery_days,customer_rating
count,483.000000,508.000000,508.000000,508.000000,508.000000,473.000000,457.000000
mean,43.954451,2.480315,658.966535,15.856299,1407.021654,5.498943,4.009847
std,18.005243,9.876836,455.886942,10.787760,6106.401935,2.884885,0.567438
min,-5.000000,-3.000000,-500.000000,0.000000,-1852.500000,1.000000,3.000000
25%,30.000000,1.000000,250.000000,5.000000,243.125000,3.000000,3.500000
50%,44.000000,1.000000,700.000000,15.000000,765.000000,6.000000,4.000000
75%,57.000000,3.000000,900.000000,25.000000,1350.000000,8.000000,4.500000
max,250.000000,200.000000,1500.000000,120.000000,98000.000000,10.000000,5.000000


In [8]:
df[df['quantity'] < 0]

,order_id,order_date,customer_id,customer_age,gender,region,product_name,category,quantity,unit_price,discount_percent,order_amount,customer_segment,is_repeat_customer,delivery_days,customer_rating
15,ORD-100015,2024-10-12,CUST-1171,49.0,Male,Kazan,Xiaomi 14,Smartphones,-3,650,5,-1852.5,Business,True,8.0,4.6


In [9]:
df[(df['customer_age'] < 0) | (df['customer_age'] > 90)]

,order_id,order_date,customer_id,customer_age,gender,region,product_name,category,quantity,unit_price,discount_percent,order_amount,customer_segment,is_repeat_customer,delivery_days,customer_rating
5,ORD-100005,2025-06-30,CUST-1186,-5.0,Male,Vladivostok,MacBook Air M3,Laptops,2,1400,5,2660.0,Business,False,9.0,3.4
10,ORD-100010,2024-07-01,CUST-1129,250.0,Female,Novosibirsk,Xiaomi TV,TVs,5,700,30,2450.0,Standard,False,3.0,4.1


In [10]:
df[df['unit_price'] < 0]

,order_id,order_date,customer_id,customer_age,gender,region,product_name,category,quantity,unit_price,discount_percent,order_amount,customer_segment,is_repeat_customer,delivery_days,customer_rating
25,ORD-100025,2024-08-03,CUST-1183,39.0,Male,Yekaterinburg,Samsung Galaxy S24,Smartphones,1,-500,20,-400.0,Business,False,1.0,3.2


In [11]:
df[(df['discount_percent'] < 0) | (df['discount_percent'] > 100)]

,order_id,order_date,customer_id,customer_age,gender,region,product_name,category,quantity,unit_price,discount_percent,order_amount,customer_segment,is_repeat_customer,delivery_days,customer_rating
20,ORD-100020,2024-03-03,CUST-1191,38.0,Male,Moscow,iPhone 15,Smartphones,2,900,120,-360.0,Business,True,1.0,4.9


In [12]:
df[df['order_amount'] < 0]

,order_id,order_date,customer_id,customer_age,gender,region,product_name,category,quantity,unit_price,discount_percent,order_amount,customer_segment,is_repeat_customer,delivery_days,customer_rating
15,ORD-100015,2024-10-12,CUST-1171,49.0,Male,Kazan,Xiaomi 14,Smartphones,-3,650,5,-1852.5,Business,True,8.0,4.6
20,ORD-100020,2024-03-03,CUST-1191,38.0,Male,Moscow,iPhone 15,Smartphones,2,900,120,-360.0,Business,True,1.0,4.9
25,ORD-100025,2024-08-03,CUST-1183,39.0,Male,Yekaterinburg,Samsung Galaxy S24,Smartphones,1,-500,20,-400.0,Business,False,1.0,3.2


In [16]:
df[(df['customer_rating'] < 0) | (df['customer_rating'] > 100)]

,order_id,order_date,customer_id,customer_age,gender,region,product_name,category,quantity,unit_price,discount_percent,order_amount,customer_segment,is_repeat_customer,delivery_days,customer_rating


### Найденные аномалии при первичном анализе датасета:

'quantity' - заказ отрицательного количества товара. 

'customer_age' - нереалистичные возрасты клиентов(-5 и 250 лет). 

'unit_price' - отрицательная цена товара. 

'discount_percent' - процент скидки больше 100% или меньше 0%. 

'order_amount' - отрицательная сумма заказа. 


Данные наблюдения требуют дальнейшей проверки и возможной очистки. 

# MISSING DATA

### В следующих столбцах были обнаружены пропущенные значения:

In [14]:
df.isnull().sum()

order_id               0
order_date             0
customer_id            0
customer_age          25
gender                15
region                 0
product_name           0
category               0
quantity               0
unit_price             0
discount_percent       0
order_amount           0
customer_segment       0
is_repeat_customer     0
delivery_days         35
customer_rating       51
dtype: int64

In [32]:
df['gender'].value_counts(dropna=False)

gender
Male      249
Female    244
NaN        15
Name: count, dtype: int64

В поле 'gender' было принято решение оставить пропущенные значения без изменений, так как пол клиента невозможно достоверно восстановить.

Была проведена проверка, что пустые ячейки уже читаются как NaN. 

In [31]:
df['customer_age'] =df['customer_age'].fillna(df['customer_age'].median())

Было принято решение заменить пропуски в столбце 'customer_age' на медиану, так как при использовании среднего значения они могут быть искажены обнаруженными выбросами, которые могут повлиять на дальнейший анализ. 

In [30]:
df['customer_rating'] = df['customer_rating'].fillna(df['customer_rating'].median())

Было принято решение заменить пропуски в столбце 'customer_rating' на медиану, так как при использовании среднего значения общее представление может испортить выброс. 

In [29]:
df['delivery_days'] = df['delivery_days'].fillna(df['delivery_days'].mean())

Было принято решение заменить пропуски в столбце 'delivery_days' на среднее значение, так как анализ не выявил выраженных выбросов в сроках доставки. В данном случае среднее значение даст большее понимание ситуации. 

In [33]:
df.isnull().sum()

order_id               0
order_date             0
customer_id            0
customer_age           0
gender                15
region                 0
product_name           0
category               0
quantity               0
unit_price             0
discount_percent       0
order_amount           0
customer_segment       0
is_repeat_customer     0
delivery_days          0
customer_rating        0
dtype: int64

После обработки пропусков значения NaN были устранены. Пропуски в gender были оставлены без изменений. 

#   POOR-QUALITY DATA

In [ ]:
df[df['quantity'] <= 0].sum()

In [ ]:
((df['customer_age'] < 0) | (df['customer_age'] > 100)).sum()

In [ ]:
(df['unit_price'] <= 0).sum()

In [ ]:
((df['discount_percent'] < 0) | (df['discount_percent'] > 100)).sum()

In [ ]:
(df['order_amount'] <= 0).sum()

Была проведена проверка, показывающая, что количество записей с некорректными данными очень мало, что говорит о случайных ошибках при заполнении, а не о некорректности всего датасета.

In [ ]:
df = df[df['quantity'] > 0]
df[df['quantity'] <= 0].sum()

In [ ]:
df = df[df['unit_price'] > 0]
df[df['unit_price'] <= 0].sum()

In [ ]:
df = df[(df['discount_percent'] >= 0) & (df['discount_percent'] <= 100)]
df[(df['discount_percent'] < 0) | (df['discount_percent'] > 100)].sum()

In [ ]:
df = df[df['order_amount'] > 0]
df[df['order_amount'] <= 0].sum()

In [ ]:
df['customer_age'] = df['customer_age'].fillna(df['customer_age'].median())
df[(df['customer_age'] < 0) | (df['customer_age'] > 90)].sum()

Чтобы избежать искажения резульатов анализа, были удалены все записи с некорректными значениями(кроме 'customer_age'), которые не могли соответствовать реальным данным. 

Некорректные записи в 'customer_age' были приведены к медианному значению, так как этот метод позволяет сохранить более реалистичное распределение возраста клиентов. 